runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -6937,16 +6937,69 @@ J9::Z::TreeEvaluator::VMifInstanceOfEvaluator(TR::Node * node, TR::CodeGenerator
    return NULL;
    }
 
+/**
+    * Generates a quick runtime test for valueType node and in case if node is of valueType, generates a branch to helper call
+    * 
+    * @param node monent or monexit node
+    * @param mergeLabel TR::LabelSymbol* to pointing to mergePoint in mainline
+    * @param callLabel TR::LabelSymbol* pointing to the helper Call Label
+    * @param cg CodeGenerator Object Ptr
+    * @return TR::Register* If Given node required a runtime test to check for valueType, it will return the register containing J9Class of given object
+    */
+static TR::Register*
+generateCheckForValueTypeMonitorEnterOrExit(TR::Node *node, TR::LabelSymbol* mergeLabel, TR::LabelSymbol *helperCallLabel, TR::CodeGenerator *cg)
+   {
+   TR::Register *objReg = cg->evaluate(node->getFirstChild());
+   TR::Register *objectClassReg = cg->allocateRegister();
+   TR::TreeEvaluator::genLoadForObjectHeadersMasked(cg, node, objectClassReg, generateS390MemoryReference(objReg, TR::Compiler->om.offsetOfObjectVftField(), cg), NULL);
+   TR_J9VMBase *fej9 = static_cast<TR_J9VMBase *>(cg->comp()->fe());
+   bool generateOOLSection = helperCallLabel == NULL;
+   if (generateOOLSection)
+      helperCallLabel = generateLabelSymbol(cg);
+   // Currently we have used a bit from the second byte in the 32-bit ClassFlags field for J9ClassIsValueType flag.
+   // To use the memory to immediate test instruction, we are generating a memory reference which accesses the second byte in the ClassFlags field.
+   static_assert(static_cast<uint32_t>(J9ClassIsValueType & 0xFF00) == static_cast<uint32_t>(J9ClassIsValueType), "Expecting J9ClassIsValueType to be within second byte of classFlags");
+   TR::MemoryReference *classFlagsMemRef = generateS390MemoryReference(objectClassReg, static_cast<uint32_t>(fej9->getOffsetOfClassFlags()) + 2, cg);
+   generateSIInstruction(cg, TR::InstOpCode::TM, node, classFlagsMemRef, J9ClassIsValueType >> 8);
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRNZ, node, helperCallLabel);
```

Comment:
This may be more on the personal preference but I'd consider adding some line breaks in between logical blocks of code. This entire function looks like one giant block of continuous code which may be hard to read. Just something to consider. I've gone ahead and improved the grammar in the comments and added line breaks to show what this function could look like:

```
static TR::Register*
generateCheckForValueTypeMonitorEnterOrExit(TR::Node *node, TR::LabelSymbol* mergeLabel, TR::LabelSymbol *helperCallLabel, TR::CodeGenerator *cg)
   {
   auto objReg = cg->evaluate(node->getFirstChild());
   auto objectClassReg = cg->allocateRegister();

   TR::TreeEvaluator::genLoadForObjectHeadersMasked(cg, node, objectClassReg, generateS390MemoryReference(objReg, TR::Compiler->om.offsetOfObjectVftField(), cg), NULL);

   // Currently the value of J9ClassIsValueType flag maps to the second (low order) byte of the class flags field. To
   // be able to use the TM instruction we need to ensure we safe guard against changes of this value since we have to
   // hardcode the offset from the class flags field to reach the byte corresponding to the bit we want to test.
   static_assert(static_cast<uint32_t>(J9ClassIsValueType & 0xFF00) == static_cast<uint32_t>(J9ClassIsValueType), 
      "Expecting J9ClassIsValueType to be within second byte of classFlags");
   auto classFlagsMemRef = generateS390MemoryReference(objectClassReg, static_cast<uint32_t>(static_cast<TR_J9VMBase *>(cg->comp()->fe())->getOffsetOfClassFlags()) + 2, cg);
   generateSIInstruction(cg, TR::InstOpCode::TM, node, classFlagsMemRef, J9ClassIsValueType >> 8);

   const bool generateOOLSection = (helperCallLabel == NULL);
   if (generateOOLSection)
      {
      helperCallLabel = generateLabelSymbol(cg);
      }

   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRNZ, node, helperCallLabel);

   // TODO: There is now the possibility of three distinct OOL sections with helper calls to be generated when
   // evaluating the TR::monent or TR::monexit nodes:
   //
   // 1. Monitor cache lookup OOL
   // 2. Lock reservation OOL
   // 3. Value types object OOL
   //
   // These distinct OOL sections may perform non-trivial logic but what they all have in common is they all have a
   // call to the same JIT helper which acts as a fall back. This complexity exists because of the way the evaluators
   // are currently architected and due to the restriction that we cannot have nested OOL code sections. Whenever
   // making future changes to these evaluators we should consider refactoring them to reduce the complexity and
   // attempt to consolidate the calls to the JIT helper so as to not have multiple copies.
   if (generateOOLSection)
      {
      TR_S390OutOfLineCodeSection *helperCallOOLSection = new (cg->trHeapMemory()) TR_S390OutOfLineCodeSection(helperCallLabel, mergeLabel, cg);
      cg->getS390OutOfLineCodeSectionList().push_front(helperCallOOLSection);
      helperCallOOLSection->swapInstructionListsWithCompilation();

      auto cursor = generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, helperCallLabel);

      TR_Debug *debugObj = cg->getDebug();
      if (debugObj)
         debugObj->addInstructionComment(cursor, "Denotes Start of OOL for ValueType Node");

      cg->getLinkage(TR_CHelper)->buildDirectDispatch(node);
      cursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, mergeLabel);

      if (debugObj)
         debugObj->addInstructionComment(cursor, "Denotes End of OOL for ValueType Node");

      helperCallOOLSection->swapInstructionListsWithCompilation();
      }

   return objectClassReg;
   }
```

Do you think it reads easier and the comments are improved?

Reply:
Thanks @fjeremic  for the suggestion. Added more spaces and also used most of suggested changes in https://github.com/eclipse/openj9/compare/c64a354e89084e22019e97161f517bd09eace94a..95e8afe940942c7102d63af15e07e73d02cb8789

Reply:
The `@return` can be simplified as well.

Reply:
Sorry, missed that, fixed it